pkg/compactor/executor.go

```diff
@@ -0,0 +1,320 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package compactor
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"time"
+
+	"github.com/failsafe-go/failsafe-go"
+	"github.com/failsafe-go/failsafe-go/retrypolicy"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/dskit/grpcutil"
+	"github.com/grafana/dskit/ring"
+	"github.com/oklog/ulid/v2"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+
+	"github.com/grafana/mimir/pkg/compactor/scheduler/schedulerpb"
+	"github.com/grafana/mimir/pkg/util"
+)
+
+// compactionExecutor defines how compaction work is executed.
+type compactionExecutor interface {
+	run(ctx context.Context, compactor *MultitenantCompactor) error
+	createShardingStrategy(enabledTenants, disabledTenants []string, ring *ring.Ring, ringLifecycler *ring.BasicLifecycler, cfgProvider ConfigProvider) shardingStrategy
+	stop() error
+}
+
+// standaloneExecutor runs compaction on a timer without external scheduling.
+type standaloneExecutor struct{}
+
+func (e *standaloneExecutor) run(ctx context.Context, c *MultitenantCompactor) error {
+
+	c.compactUsers(ctx)
+
+	ticker := time.NewTicker(util.DurationWithJitter(c.compactorCfg.CompactionInterval, 0.05))
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			c.compactUsers(ctx)
+		case <-ctx.Done():
+			return nil
+		case err := <-c.ringSubservicesWatcher.Chan():
+			return errors.Wrap(err, "compactor subservice failed")
+		}
+	}
+}
+
+func (e *standaloneExecutor) stop() error {
+	return nil
+}
+
+func (e *standaloneExecutor) createShardingStrategy(enabledTenants, disabledTenants []string, ring *ring.Ring, ringLifecycler *ring.BasicLifecycler, cfgProvider ConfigProvider) shardingStrategy {
+	allowedTenants := util.NewAllowList(enabledTenants, disabledTenants)
+	return newSplitAndMergeShardingStrategy(allowedTenants, ring, ringLifecycler, cfgProvider)
```

Comment:
The contents of the `newSplitAndMergeShardingStrategy` could probably move here now, since it presumably won't be used anywhere else? This would also be symmetrical with the schedulerExecutor.createShardingStrategy